src/Codec/JVM/ASM/Code/Instr.hs

```diff
@@ -49,18 +53,21 @@ instance Functor InstrM where
   fmap = liftM
 
 instance Applicative InstrM where
-  pure = return
+  pure x = InstrM $ \_ s -> (# x, s #)
   (<*>) = ap
 
 instance Monad InstrM where
-  return x = InstrM $ \_ s -> (# x, s #)
+  return = pure
   (InstrM m) >>= f =
     InstrM $ \e s ->
       case m e s of
         (# x, s' #) ->
           case runInstrM (f x) e s' of
             (# x', s'' #) -> (# x', s'' #)
 
+instance MonadFail InstrM where
+  fail = error
```

Comment:
There are a bunch of places with partial pattern matches in the code, but I coudn't immediately see a better implementation.

So this will come into play a bit later. But it should probably be updated with a better error message and wired into the older Monad instance's fail as well.